File: client/lib/schema/measurement_history.dart
import 'package:client/schema/device.dart';
import 'package:client/schema/food.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'measurement_history.freezed.dart';
part 'measurement_history.g.dart';

@freezed
class MeasurementHistory with _$MeasurementHistory {
  const factory MeasurementHistory({
    required int id,
    @JsonKey(name: "food_id") required String foodId,
    required double weight,
    String? deviceId,
    Device? device,
    Food? food,
    @JsonKey(name: "created_at") DateTime? createdAt,
  }) = _MeasurementHistory;

  factory MeasurementHistory.fromJson(Map<String, dynamic> json) => _$MeasurementHistoryFromJson(json);
}

@freezed
class RecordHistoryRequest with _$RecordHistoryRequest {
  const factory RecordHistoryRequest({
    @JsonKey(name: 'device_token') required String deviceToken,
    @JsonKey(name: 'nfc_uid') required String nfcUid,
    required double weight,
  }) = _RecordHistoryRequest;

  factory RecordHistoryRequest.fromJson(Map<String, dynamic> json) => _$RecordHistoryRequestFromJson(json);
}

File: core-api/pkg/handler/schema/measurement_history.go
package schema

import (
	"time"

	"github.com/gin-gonic/gin"
)

type MeasurementHistory struct {
	Id        int64     `json:"id"`
	FoodId    string    `json:"food_id"`
	Weight    float32   `json:"weight"`
	DeviceId  *string   `json:"device_id"`
	Device    *Device   `json:"device"`
	Food      *Food     `json:"food"`
	CreatedAt time.Time `json:"created_at"`
}

type RecordHistoryRequest struct {
	DeviceToken string  `json:"device_token"`
	NfcUid      string  `json:"nfc_uid"`
	Weight      float32 `json:"weight"`
}

type CreateMeasurementHistoryRequest struct {
	Weight   float32    `json:"weight"`
	RecordAt *time.Time `json:"record_at"`
}

type CreateMeasurementHistoryFromAppRequest struct {
	Weight   float32 `json:"weight"`
	NfcUid   string  `json:"nfc_uid"`
	DeviceId *string `json:"device_id"`
}

type MeasurementHistoryController interface {
	RecordHistory(c *gin.Context)
	FindHistory(c *gin.Context)
	CreateHistory(c *gin.Context)
	CreateHistoryFromApp(c *gin.Context)
}
